### PR TITLE
Added types for list-stream

### PR DESCRIPTION
**Pull request by IanStorm**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/list-stream/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// Type definitions for list-stream 1.0
+// Project: https://github.com/rvagg/list-stream
+// Definitions by: IanStorm <https://github.com/IanStorm>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/// <reference types="node" />
+
+import { Duplex, DuplexOptions } from "stream";
+
+declare namespace ListStream {
```
Comment (uniqueiniquity) [inline types/list-stream/index.d.ts]: I don't think you need this `namespace`.

**Review by uniqueiniquity — CHANGES_REQUESTED** [review]

File: types/list-stream/index.d.ts (hunk 1 above)
Comment (IanStorm) [reply types/list-stream/index.d.ts]: @uniqueiniquity Thanks for checking my Pull Request, and thanks for the feedback.

While trying to fix that issue, I actually came a cross some bigger mistake that I have done while creating the definition. However, with the latest commit (f6a44f54c73a0b255e7b36a1953824e518165996) the definition should be proper. Sorry that it is a major change.

File: types/list-stream/index.d.ts [hunk 2]
```diff
@@ -7,16 +7,21 @@
...
-declare namespace ListStream {
+interface ListStreamMethod {
+    (callback?: (err: Error, data: any[]) => void): ListStream;
+    (options?: DuplexOptions, callback?: (err: Error, data: any[]) => void): ListStream;
 }
 
-declare class ListStream extends Duplex {
-    constructor(callback?: (err: Error, data: any[]) => void);
-    constructor(options?: DuplexOptions, callback?: (err: Error, data: any[]) => void);
+interface ListStreamConstructor extends ListStreamMethod {
```
Comment (uniqueiniquity) [inline types/list-stream/index.d.ts]: @IanStorm  I'm surprised you moved away from using a class here. Was there a particular reason that you thought a new-able constant would be better?

Comment (uniqueiniquity) [reply types/list-stream/index.d.ts]: I see, that makes sense!

I do have a recommendation regarding your "import style" point.
Under the ES6 spec, it is not correct to expect something imported with `import * as foo from "bar";` to be callable or newable. If consumers of this package want to use ES6-style imports in TypeScript, they should use `import ListStream from "list-stream";` and the `--esModuleInterop` compilation flag. You can do the same in your tests by adding that compilation option to the tsconfig.json in this folder.

See [here](https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#support-for-import-d-from-cjs-form-commonjs-modules-with---esmoduleinterop) and [here](https://github.com/definitelyTyped/DefinitelyTyped/#should-i-add-an-empty-namespace-to-a-package-that-doesnt-export-a-module-to-use-es6-style-imports) for more information.

File: types/list-stream/list-stream-tests.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import * as ListStream from "list-stream";
```
Comment (uniqueiniquity) [inline types/list-stream/list-stream-tests.ts]: See previous comment.

**Review by uniqueiniquity — CHANGES_REQUESTED** [review]

File: types/list-stream/index.d.ts (hunk 2 above)
Comment (IanStorm) [reply types/list-stream/index.d.ts]: Thank you for the recommendation. The links you provided were really helpful, not only with respect to this merge request, but also to improve my general understanding. :)
I pushed new commits (f368a2dbfdd73511fbde3ab961ec111ce0486646 and 3f185078d8bdf57c6669b08a5ea135de65555d9d) to include the required changes.

Keeping in mind that my goal was to support "import style" and "require style" I was thinking that the "tests" file should also reflect that. Currently it only reflects the "import style". So what would you say about having two `list-stream-tests.ts` files, i.e. one for "import style" renamed to (for example) `list-stream-tests-import.ts`, and another for "require style" named (for example) `list-stream-tests-require.ts`? Do you think it is a good idea? Is there any convention on how to name or create the test files?

File: types/list-stream/list-stream-tests.ts (hunk 3 above)
Comment (IanStorm) [reply types/list-stream/list-stream-tests.ts]: Should be resolved by https://github.com/DefinitelyTyped/DefinitelyTyped/pull/30669#discussion_r238110651.

File: types/list-stream/index.d.ts (hunk 2 above)
Comment (uniqueiniquity) [reply types/list-stream/index.d.ts]: There definitely does need to be a `list-stream-tests.ts` file. However, you can make a `tests` directory and add your two test files there, then import them into `list-stream-tests.ts`. Alternatively, you can use both imports in a single file:
```ts
import ListStream = require("list-stream");
import ListStream2 from "list-stream";
import { obj } from "list-stream";
```

**Review by uniqueiniquity — APPROVED** [review]

**Comment by uniqueiniquity** [comment]
Waiting to merge until mentioned test changes are made.

**Comment by IanStorm** [comment]
> Waiting to merge until mentioned test changes are made.

@uniqueiniquity I made the changes. Thanks for waiting.

> There definitely does need to be a `list-stream-tests.ts` file. However, you can make a `tests` directory and add your two test files there, then import them into `list-stream-tests.ts`. Alternatively, you can use both imports in a single file:
> 
> ```ts
> import ListStream = require("list-stream");
> import ListStream2 from "list-stream";
> import { obj } from "list-stream";
> ```

I considered both of the options you suggested and decided to create separate files.

**Comment by uniqueiniquity** [comment]
Congratulations on your first DefinitelyTyped contribution!
Thank you for being a part of the community!

**Pull request merged**